### DISCUSSION TITLE
[$200+ Bounty] Integrate/Bundle Level Designer Toolkit (LDtk) into GDevelop as a tilemap editor

### DISCUSSION THREAD

**Discussion by blurymind** [4ian/GDevelop#2352] [Enhancement ideas (and their technical discussions)]
## Read the discussion for full context, but this issue will probably require to start with:

- [ ] Support for multi-layer tilemaps
- [ ] Support for LDtk JSON format (without having to export to Tiled)
- [ ] Support for reloading tilemaps when the file is changed

All of these are useful on their own (and can lead to some bounty if agreed beforehand). Integration of the full LDtk editor might be more tricky and need a good design/technical solution.

---

Led is a level editor, that is similar to tiled- but with the exception that  it can run inside electron. What is even more exciting is how clean and simple the interface is, while being capable of very powerful autotiling
https://www.youtube.com/watch?v=pYBWGt8wbsU
https://deepnight.net/tools/ldtk-2d-level-editor/
https://github.com/deepnight/ldtk

ticket on the ldtk tracker
https://github.com/deepnight/ldtk/issues/126

it has powerful auto-tiling, but it can also be  used in the traditional way without it
![fullLevelEdit](https://user-images.githubusercontent.com/6495061/94418415-4aa61100-0179-11eb-9aee-0451fe67b296.gif)

This makes me wonder if we can bundle it like we bundle piskel. Still needs to be researched a bit :)

Trello board card:
https://trello.com/c/yCXSSFFu/456-integrate-led-level-editor-inside-gdevelop-for-designing-tilemaps-levels-with-it

**Comment by 4ian** [comment]
It seems very nicely done :) Seems like the output format is custom (not the one from Tiled) so this will need an adapter I think.
Overall yes I think it would be a great addition - though we'll have to see how to make it seamlessly integrated (or as integrated as possible) to the existing scene editor.

I've created a card here: https://trello.com/c/yCXSSFFu/456-integrate-led-level-editor-inside-gdevelop-for-designing-tilemaps-levels-with-it

**Comment by Silver-Streak** [comment]
@4ian funnily enough, I've been using LEd since its newest update and asked the dev if they could output to the standard Tiled Json format.

They actually added it to their roadmap for 0.3.0.

In theory this would make it compatible with the work @blurymind 
 is doing.

**Comment by 4ian** [comment]
Awesome! Making even more interesting to get this tilemap object working :) So we can start using it in games and see what's the best way to integrate an advanced editor like this.

**Comment by blurymind** [comment]
if we could embed an external editor directly in gdevelop - as a tab, rather than a modal window - that could make it feel more integrated :) in any case if the developers make a tiled importer/exporter, it  will save us the effort to write an LED importer

**Comment by Silver-Streak** [comment]
Just a heads up here for @blurymind and @4ian 

LE-d has hit 1.0 and renamed to Level Designer Toolkit: https://deepnight.net/tools/ldtk-2d-level-editor/

**Comment by blurymind** [comment]
I am thinking that the best way forward for this is to add the ability to parse LdTk JSON files to the tilemap extension.
The file data can be used to automatically detect if ldtk or tiled was used to create it - by the helper class.

From there on it can be sent to the correct parsing method.

Problems to overcome before ldtk can de embedded as an external editor:
- the concept of resource dependencies is missing in gdevelop. ldtk can have multiple tilemaps, each using multiple tilesets with  multiple atlas images. All of their relative paths are in the ldtk json data, however gdevelop needs them all to be automatically added as new resources with the resource manager if they arent. Since we dont want the user to be manually selecting bazillion files which would lead to bad UX, gdevelop's tilemap extension needs to be able to handle this elegantly, just using the data it received from ldtk.
- a new parser for the ldtk json data structure needs to be added to the helper class of the tilemap extension


Potential alternative way to include LDTK in gdevelop is to make a new react component which wraps it - that way we could put it directly inside gdevelop as a new tab type and it will also work in the web version. Ldtk as a react component can be developed on a different gihub repository and added to npm as a module. This can attract more maintainers to it.

**Comment by Bouh** [comment]
> Potential alternative way to include LDTK in gdevelop is to make a new react component which wraps it - that way we could put it directly inside gdevelop as a new tab type and it will also work in the web version. Ldtk as a react component can be developed on a different gihub repository and added to npm as a module. This can attract more maintainers to it.

I agree.
Piskel was cool but open up in a new window is limited and break the immersion, i'am not fan.
With full integration of Ldtk in a new tab it will make this integration a lot better and multiplatform! Yay!

**Comment by Silver-Streak** [comment]
So, I've also posted over on the Pixi-tilemap Github about the possibility of adding LDtk formatted JSON support. I'm waiting for clarification from the maintainer as I'm not quite understanding his question.

Assuming that doesn't go through, if I understand correctly from @blurymind , this idea needs 3 things:
1. Wrap LDTk in a react component so it can show up as a tab in GD5.
2. Build a parser for LDtk JSON format to allow Pixi-Tilemap to ingest the tilemaps
3. Add Resource Dependencies or a way for GDevelop to properly understand the data/image sources from the LDtk json

Is the above accurate?
Is there an alternate for number 2? (Such as doing something that transcribes Tiled TMX to JSON since LDtk can already output to that format?)
Is there an alternate for number 3? (Such as instructing users that multiple-tilesets are not currently supported?, or something in the tileset object that shows the related resources?)

If the above are all accurate, is @4ian okay with me opening up a bounty on this?

**Comment by 4ian** [comment]
> Wrap LDTk in a react component so it can show up as a tab in GD5.

This is more or less the ideal situation yes. Unsure how doable it is, because it would need also disabling a bunch of LDtk feature that we would not need (some configuration settings or these kind of stuff that would not be relevant for GDevelop).

> Is there an alternate for number 2? (Such as doing something that transcribes Tiled TMX to JSON since LDtk can already output to that format?)

I think we could as well handle directly as you said the LDtk JSON format, because it's not that complex and may be better than triggering a translation to Tiled TMX, then to Tiled JSON, all of this to finally be able to open it.

> Is there an alternate for number 3? (Such as instructing users that multiple-tilesets are not currently supported?, or something in the tileset object that shows the related resources?)

In a first version, we can add some limitation to the LDtk supported features. Like one tileset, or separately export tilemaps maybe.

> If the above are all accurate, is @4ian okay with me opening up a bounty on this?

Feel free to do so! Note that this kind of task seem actually more complex than actually adding the initial Tilemap object that we currently have (so that everyone has an expectation on the complexity of this).

**Comment by Silver-Streak** [comment]
Hey all, while it's not as large as I'd like it to be right now, I've started the bounty for this effort here: https://www.bountysource.com/issues/93298046-idea-add-new-externral-editor-to-edit-tilemaps-ldtk

**Comment by Silver-Streak** [comment]
@4ian, @Bouh, or @blurymind  (or someone who can): Can we rename this to "Idea: Integrate/Bundle Level Designer Toolkit (LDtk) into GDevelop as a tilemap editor" or something similar? Just want to clean up the title for the bounty.

**Comment by 4ian** [comment]
Updated, thanks! :)

**Comment by blurymind** [comment]
I'd say lets investigate with the ldtk developer if we can get it as a react component. He might be willing to assist in some way or even engage this feature, since its marked as a milestone feature on the tracker there
https://github.com/deepnight/ldtk/issues/126

**Comment by Silver-Streak** [comment]
General heads up, I was able to bump this up to roughly where I wanted to start it at. The bounty is now at $150: https://www.bountysource.com/issues/93298046-idea-integrate-bundle-level-designer-toolkit-ldtk-into-gdevelop-as-a-tilemap-editor

**Comment by Silver-Streak** [comment]
@4ian as a heads up, I think making this a discussion may break the bounty on it at bountysource.

**Reply by 4ian** [reply]
Ah right, let's see what happens, we might do a new issue to advertise this.

**Reply by 4ian** [reply]
I think it's fine, the bounty is still there and linked to the issue... which redirects to this discussion?
It might make it harder to find it, but for this I'm thinking of a "Good First Contribution" category, so that it's easy to find them.

**Reply by Silver-Streak** [reply]
@4ian  Unfortunately, this did break the bounty. It's showing closed now and someone can try to claim the bounty.

Should we reopen the issue?

Edit: It also looks like Bounties can't be posted on discussions at all. Might be worthwhile to leave issues open with some sort of "please keep all conversation to this discussion" with a link to the discussion, for bounty purposes?

Double edit: I've also reached out to the Bountysource support group to ask for best processes around these as well, going forward.

**Reply by 4ian** [reply]
I unfortunately can not re-open the issue :( 
EDIT: I'll reach out to bountysource to ask what to do.

>  Might be worthwhile to leave issues open with some sort of "please keep all conversation to this discussion" with a link to the discussion, for bounty purposes?

Yes I guess.

> I've also reached out to the Bountysource support group to ask for best processes around these as well, going forward.

Yes, they are a bit lacking behind GitHub in terms of features - and it's even hard to find on their website a list of all the bounties.. 

Might be worth keeping these kind of issues with bounties open (or make an issue specifically when someone wants to post a bounty), but ideally the process would be simpler. By the way, are there alternative to bountysource? We could investigate other ways to advertise and support issues/feature requests with an easy way to put bounties on it.

**Reply by Silver-Streak** [reply]
> By the way, are there alternative to bountysource? We could investigate other ways to advertise and support issues/feature requests with an easy way to put bounties on it.

I had done some digging, nearly every other bounty service I could find only accepted bounties in, or paid bounties out, in cryptocurrency, which is relatively unappealing for people wanting to contribute based off how small most of the bounties were on those platforms.

Absolutely happy to use another service if there's a better one out there.

I did hear back from the Bountysource people, they're going to try to reopen the current bounty, and are actively having discussions about Discussion bounties.

**Comment by Silver-Streak** [comment]
The bounty has been re-opened, and I've boosted it up to $200

https://www.bountysource.com/issues/93298046-idea-integrate-bundle-level-designer-toolkit-ldtk-into-gdevelop-as-a-tilemap-editor

**Comment by Silver-Streak** [comment]
@blurymind 4ian and I had a small discussion around this, and I had some questions:

- Do you think it'd be more feasible for us to change the bounty to be around updating the tilemap object to also support LDtk json files?
- Do you know if it's possible to support multi-layer tilemaps as multiple layers in the scene? Is that something that could be treated as a separate bounty?  Or is there a fundamental issue with pixi-tilemap that would make that not possible? 

Edit: Not sure if it helps or hinders any of the above, but they did release a Javascript parser for their JSON: https://ldtk.io/api/ and https://ldtk.io/files/quicktype/LdtkJson.js

**Reply by 4ian** [reply]
I've at least listed in the first message of the discussion the requirements/sub tasks that would be useful to do first. Happy to consider some as a separate issue, or even make a bounty for each of those, as they would be valuable on their own.

**Reply by Silver-Streak** [reply]
That makes sense, I'd love to get @blurymind 's input on these.

I think LDtk json (or .ldtk) support would be the most helpful (and since they have a JS parser, maybe the easier of the items), but supporting multi-layer tilemaps as actual multiple layers (for parallax and other uses) could be beneficial as well.

I feel like reloading tilemaps when the JSON is updated may only be useful as a workaround for not-integrating a map editor directly.

Obviously I'm open to challenge on all of this as these are just _feelings_ and I'm no expert. I'd _really_ love to get an integrated tilemap editor, so if it makes more sense to split these out as separate bounties too, I'm fully in support of that as well.

Edit: To clarify, I think LDtk json/.ldtk support is the most helpful explicitly because we've seen a lot of users on Discord get confused about using Tiled Map Editor. LDtk has a much more modern, easier to use, and faster interface than Tiled Map Maker, I feel like newer/less technical users would have a better overall experience. In a lot of cases, I feel that it is easier to use LDtk to make the base map, import it into Tiled, then export as Tiled JSON than it is to make the map in Tiled to begin with.

Separately, LDtk file support could eventually (in a much later bounty/PR) help with using LDtk's intgrid, enum grid, and point maps/arrays as actual game logic items. Point arrays being usable in the engine would help a ton of people with building out enemy paths or other movement based items, Intgrid layer support could help people build out collisions or other movement based logic easier, etc.

**Comment by blurymind** [comment]
I would love to have these split into different tasks/bounties as they are each a feature of their own that is a step towards integrating ldtk into gdevelop's ide.

- parsing the ldtk json format - while it has a json parser, we still need to write something to feed the data to pixi-tilemap. As I mentioned before the helper class will need to be updated to detect if the json file was made with ldtk and then execute a different method to build the tilemap and the tileset. The parsing itself may come with the limitation of only being able to handle one atlas per one tileset for the tilemap, but we may come up with a way to overcome this by introducing some sort of a resource dependency system or a way for the extension to automatically add dependent resources found in the relative paths listed inside the starting json file.

`Do you know if it's possible to support multi-layer tilemaps as multiple layers in the scene? `
multilayer tilemaps are already supported for tiled json files, so they wont be a problem imo to add as part of the parsing task. Doesnt require a separate task

- Automatically reload a json (and maybe other) resources when they have been edited outside of gdevelop - I imagine the electron version will need some sort of a watch file update mechanic for resources to achieve that.
- Bundling ldtk with gdevelop - this will require at least the parsing part to be implemented first and is probably the biggest challenge to do. It will require to run either as a modal external editor or come up with a more integrated way. Also we will need to write a middle layer for gdevelop to communicate with it and vice versa

I think splitting this into different tasks with bounties is going to honestly make it much more approachable for myself or others to pick up.

So my advice is to keep this issue as a task to bundle ldtk, but create other issues for the parsing of ldtk json format and automatic file reloading as they would be beneficial to have on their own and are complex enough to do in separate prs

**Reply by Silver-Streak** [reply]
> multilayer tilemaps are already supported for tiled json files, so they wont be a problem imo to add as part of the parsing task. Doesnt require a separate task

@blurymind I meant more like treating layers in tilemaps as individual layers in GDevelop, so you can do parallaxing of a background layer. Unless that's possible and I'm just not seeing it in the event system?

@4ian  Based off blurymind's info, it seems like the next biggest step would be parsing the ldtk format. Do we want to open up a separate issue for that, for bounty purposes?

**Reply by Silver-Streak** [reply]
4ian and I talked on this, and I'm going to write up an issue for the ldtk parsing, so we can open a bounty on it.

**Reply by Silver-Streak** [reply]
https://github.com/4ian/GDevelop/issues/2434 has been opened for the ldtk map parser.

**Comment by Silver-Streak** [comment]
@blurymind As a heads up, looks like a new version of pixi tilemap is out. Sukant pinged me about it on https://github.com/pixijs/tilemap/issues/109#issuecomment-798961170

Not sure if it's something that can help with stuff if we update.

Edit: Oops, he pinged you too. 😄  Either way, thought I'd chime in here.

**Comment by willianholtz** [comment]
Hey @blurymind, how is the development of this implementation? What about Tiled collisions? Any updates?

**Comment by blurymind** [comment]
Hi, it is still going although a bit slower that I would like to. The pixi6
upgrade issue got me stuck. I might raise it on the pixi-tilemap tracker.

Tiled collisions is something that should be addressed in a separate
task.It is not something that is currently worked on

**Comment by Silver-Streak** [comment]
Thanks for the update @blurymind 

I agree that Tiled Collisions will likely need to be a separate PR/bounty/etc. Honestly I think Intgrid and enums in LDtk are a much bigger boon than the collisions from Tiled at this point. With Tiled Collisions, you only save a few steps of drawing out your collisions with a single object.

With Enums and IntGrids, if we get the logic/events added around them,  we could literally build out things like enemy paths, spawn points, checkpoints, etc.

Obviously this is all my opinion, but something I'd likely put out another bounty on someday.

**Comment by Silver-Streak** [comment]
@4ian The Tilemap Editor bounty has closed out on Bountysource again. I've reached out to ask for next best steps to get it reopened, but you might need to just claim the bounty and we can reopen a "Longer-term" issue on this.

**Reply by 4ian** [reply]
Can you share the link?

**Reply by Silver-Streak** [reply]
https://www.bountysource.com/issues/93298046-idea-integrate-bundle-level-designer-toolkit-ldtk-into-gdevelop-as-a-tilemap-editor

**Reply by Silver-Streak** [reply]
I haven't heard back from the bountysource support on this. I'm betting we probably need to claim the bounty and open a new issue on it so the bounty stays open.

**Reply by Bouh** [reply]
The management and support on bountysource is really not aweful, support don't respond, the delays are extremely long. Is there still someone working behind bountysource?
My +200$ cashout is still pending.

We should really change ours process and avoid bountysource, it doesn't feel good...

**Reply by Silver-Streak** [reply]
I usually get a response in a few days. 

@Bouh Wait, you haven't received the bitmap font payout? It says it's already paid out. I can escalate if so.

**Reply by Bouh** [reply]
The cash is approved and is in my profil on the website, only the cash out to Paypal is still pending.
In general bountysource take weeksssss to unlock any movements.

**Reply by 4ian** [reply]
Very sad it's taking so much time :/ Let me know and @Silver-Streak if it's paid in a few more days. The bountysource CEO contacted me for an interview to know more about our usage of it, so this certainly is an important feedback.

This being said, I think we should consider rolling our own solution at some point, that could be tied to a GDevelop account, and that could be used for core contributions, but also for extensions/examples/assets. At least we could make the payment instant 😬

**Reply by Silver-Streak** [reply]
@Bouh I've gotten confirmation back from support that they're looking into it. We'll see what happens.

@4ian I definitely agree we should dig into alternative bounty hosts if you don't make your own, I've been digging into options on the discord, but I'll end this side-conversation here since it's not relevant to this issue.

**Comment by Silver-Streak** [comment]
@blurymind @4ian  The bountysource funds have finally been transferred, and I'm now able to open the bounty on Rysolv. 

- Github Issue is here: https://github.com/4ian/GDevelop/issues/2991
- Rysolv bounty is here: https://rysolv.com/issues/detail/d62c42cc-1a30-445a-a0ed-e19a62478005

**Reply by 4ian** [reply]
Great to hear!

**Comment by blurymind** [comment]
Thats fantastic news! Thank you for sorting this out
